### DISCUSSION TITLE
Bevy UI: High level targets for Widgets

### DISCUSSION THREAD

**Discussion by Weibye** [bevyengine/bevy#5604] [Ideas]
In order to improve the native UI experience in Bevy and seriously begin tackling #254, we need to have a solid idea of where we are going and somewhat of an idea of how it needs to look like once we get there. 

`bevy_ui`, in its current state, is excessively verbose and cumbersome to work with. We both need to improve the ergonomics of the lower level abstractions and develop higher level abstractions. 

In order to break ground on one of those higher level abstractions, I want to start a discussion on the targets and requirements for **Widgets** in Bevy. 

### Dictionary
- UI: User interface. Abstraction layer responsible for allowing the user to interact with the application or game.
- Node: The graphical user interface is constructed of a hierarchy of containers which are stored as nodes in a graph, hence they are often referred to as nodes.
- Style: A collection of parameters that define the look and feel of the node (size, border, margin, padding, colors).
- Theme: A higher level abstraction that take a collection of styles and apply them across multiple nodes. An example of its use would be light, dark and high-contrast themes. The feature should cover how to define themes as well as how to apply them. (This feature has yet to be conceptualized and is no more than a vague idea at this point in time.)
- Immediate mode: The UI is cleared and constructed anew every single frame.
- Retained mode: The UI is only redrawn when something changes.
	- `bevy_ui` is a retained-mode UI framework.

## Widget definition and requirements

1.  Widgets are the logical building blocks of the UI
    1.  Users should be focused on composing widgets together in order to build UIs.
    2.  While the underlying mechanics should be discoverable, they should not be in the foreground when building UIs.
2.  Widgets contain predefined behavior
    1.  How they respond to various user input
    2.  How they interact with UI navigation & focus systems
    3.  How they send state to the ECS
    4.  How they fetch state / data from the ECS
3.  Widgets have a predefined look and feel
    1.  Users can change the visual style of individual widgets to suit their needs
        1.  Fonts
        2.  Icons
        3.  Colors
        4.  Visual decorations
        5.  Size
        6.  Layout
    2.  Users can define themes which are applied to entire classes of widgets at once.
        1.  Themes can be shared across projects and _mostly_ work.
        2.  Themes can be locally overridden for specific widgets.
        4.  It is easy to understand the complete set of themes being applied to a widget and how they interact.
        5.  Colorblind, high-contrast and light/dark mode themes are easy to define.
4. Widgets are animatable
    1. Animating UI should be the same as animating any other property in the engine
5.  Widgets are reusable
    1.  Users can define new, custom widgets in the same fashion that `bevy_ui` does.
    2.  Users can construct composite widgets by combining multiple existing widgets.
    3.  Variations on widgets can be easily constructed based on the properties you want to change.
    4.  Widgets can be defined in code as well as via deserialization from an asset.
6.  Widget trees are navigable
    1.  Using mouse
    2.  Using keyboards
    3.  Using multitouch
    4.  Using gamepad
    5.  Using screen readers
    7.  Widget trees are navigable by default
    8.  The exact way in which UIs can be navigated can be finely controlled by the programmer
7.  Widget layout is flexible
    1.  Creating standard rectangular screen-space layouts with widgets should involve as little boilerplate as possible.
    2.  UI widgets can be defined in either screen space or world-space coordinates
    3.  Layouts can be composed by mixing high-level widgets with low level primitives such as rectangles or hexagons.
    4.  The scale of the entire UI, classes of widgets or specific widgets can be increased or decreased gracefully, to support diverse screen resolutions, user needs and localization.
8.  Widgets can be constructed in a number of ways
    1.  Directly using the underlying primitives
    2.  In code, using a high-level abstraction
    3.  In a data-driven way from a serialized format
9.  Essential supplementary features to consider
    1.  Text localization
    2.  Left-to-right and right-to-left writing modes (bidirectional text)

[Here is a list of Widgets](https://github.com/bevyengine/bevy/issues/1953) we would like to support eventually.

### Note
These requirements are aspirational. It could be that some of these requirements turn out to be mutually exclusive or even impossible within our tech-stack, in which case we'll come back to this and re-evaluate. 

We do however need a solid goal to aim for.

### Prior Art & Relevant reading material
- [RFC #1: UI Building Blocks and Styles](https://github.com/bevyengine/rfcs/pull/1)
	- Strongly recommend reading through this to get a sense of what other thoughts have been done on the subject.
	- Contains good arguments for why we need a native `bevy_ui`.
	- Leans heavily towards the problem of styling, style-propagation and theming. I think we need a more clear idea of what *Widgets* should be before ready to dive into that design.
- [RFC #22: UI Central Dogma and A High-Level Compositional Design](https://github.com/bevyengine/rfcs/pull/22)
	- The outcome of this discussion should be fed into that RFC, and from there we might be able to find new action-points.
- [High-level UI philosophy](https://github.com/bevyengine/bevy/discussions/1996)
	- Describing UI as a pure function of state
	- Inspiration from Elm & React
	- Contains good learnings we might want to consider
- [Bevy in production: making reusable widgets with bevy and egui](https://github.com/bevyengine/bevy/discussions/5522)
	- @aevyrie outline how they (Foresight) define widgets, building on top of `bevy_egui`
	- Example of how to use the widget in the layout: 
		- `widget::<MyCoolWidget>(world, ui, "unique name");`
- @TheRawMeatball's `ui4` exploration: [Details here](https://github.com/TheRawMeatball/ui4)

### Tools and ideas we might need to build this
- Prefabs
	- Defines a group of entities that are connected by relations (parent/child ++)
	- Ideally, contain a definition of what components these entities should have
	- Ideally, contain a definition of what data should go into these components
	- APIs for spawning and setting up everything in one go
	- (Idea in early stages of development, subject to change)
	- [Discord discussion on the topic](https://discord.com/channels/691052431525675048/996172755600949348/996172757110902794)
- Event Bubbling
	- @aevyrie's idea: of propagating events up the node-hierarchy in order to achieve inter-widget communication
	- [Details of how it looks in use in this example](https://github.com/aevyrie/bevy_mod_picking/pull/137/files#diff-71beb1c3083de60af10825b10bd69789554bde153bb8b6346c1237571ae4ba88)
- [RFC #41: Generic UI navigation system](https://github.com/bevyengine/rfcs/pull/41)
	- Implementation here: #5378
- Archetype Invariants: #1481
- Relations: #3742
- One-shot systems: #4090
- Kinded entities: #1634

### User Stories and Experiences
- [@alice-i-cecile's experience of building a Sudoku game in Bevy 0.5](https://github.com/bevyengine/bevy/discussions/2235)
	- Outlines a lot of good reasons why we need a higher level abstraction

## Questions to answer

- Do you agree with the requirements outlined above?
	- Are there adjustments you think are important to make?
- What should the concrete definition of a Widget be, in terms of `entities`, `components`, `systems` and `events`?
- How should the user define or declare a widget?
- How should the user create instances of the widget when building layouts?
- How should widgets be composable?
- How should widgets be stylable, themable?
- How should the user hook up communication between widgets?
- How should widgets send state to the world?
- How should widgets request state from the world?
- Are there other important questions that should be added to this list?

**Comment by alice-i-cecile** [comment]
First crack at these questions!

> Do you agree with the requirements outlined above?

I support all of the requirements outlined above. I think they may be clearer if grouped though.

> Are there adjustments you think are important to make?

I think that UI scaling is another important accessibility tool that we should be careful to design around.

I also think that #4090 belongs under the list of tools we might need.

> What should the concrete definition of a Widget be, in terms of entities, components, systems and events?

A widget should be a collection of one or more entities (and their components), connected via relations (to start with, the Parent/Children components).

> How should the user define or declare a widget?

Widgets should be declared using a multi-entity variation of the `#[derive(Bundle)]` invocation that I'll call a "prefab".

Builder methods can be added to allow the generation of variants that share the same structure (entities and components) but have different configuration.

> How should the user create instances of the widget when building layouts?

There should be three methods:

1. Loaded as part of scenes.
2. Spawned via commands
3. Spawned via direct world access

This is consistent with the existing paradigms for ECS, and allows us to use standard Rust tooling like builder methods.

> How should widgets be composable?

Widgets should be nestable, via hierarchical entity composition.

This can either be done ad-hoc, or by constructing a new prefab type, as outlined above. 

> How should widgets be stylable, themable?

Each widget should have at least one defining marker component, in addition to the default `Widget` marker component.

These marker components can be combined with `Added` query filters to apply styling to each new widget before they are rendered (with no command flushes in between).

Themes are stored in a central resource and are applied in a well-defined order on the basis of the marker components present on each entity.

> How should the user hook up communication between widgets?

Through event bubbling.

> How should widgets send state to the world?

Through one-shot systems or event emission.

> How should widgets request state from the world?

Via ordinary systems.

> Are there other important questions that should be added to this list?

Defining selection strategies for widgets and widget sub-components based on non-local properties seems quite hard until we get very advanced relations features.

For example on the web it's common to want to style all elements that are the child of an element that has a certain property. Users would be able to hand-write systems to search for this case and add/remove the appropriate marker components, but that will be tricky and we should start thinking about ways to improve this in both the short and long term.

**Reply by Weibye** [reply]
> I support all of the requirements outlined above. I think they may be clearer if grouped though.

Any recommendation to the grouping, or how that should look?

**Reply by alice-i-cecile** [reply]
My take on it is here: https://hackmd.io/@1w5KSqtVQSqQ-xUpP2tRzw/S1hv9v66c/edit

**Reply by Weibye** [reply]
Updated the post with your, more clear, definitions!

**Reply by Weibye** [reply]
Added one-shot systems and kinded entities

**Comment by afonsolage** [comment]
A couple days ago I started my adventure in building [my own Widgets](https://github.com/afonsolage/bevy_ui_proto_widgets) using pure `bevy_ui`. It is still very early stage in development, but I have some thoughts to share. 

> Do you agree with the requirements outlined above?

Sure, but I don't think Bevy should have it's own common widgets, like stated on https://github.com/bevyengine/bevy/issues/1953. It could have some basic widgets, like buttons and maybe other 2 or 3, because UI is part of the game look'n'feel so users will have to make their own widgets anyways (unless it's a prototype or something like that), so Bevy should focus on ways of allowing to create new widgets as easier as possible so users may create hundreds of crates with custom themes and widgets (like RTS, MMORPG, FPS and so on).

> Are there adjustments you think are important to make?

Everything that is missing, is already address somewhere else, be it RFC, PR or another proto crate.

> What should the concrete definition of a Widget be, in terms of entities, components, systems and events?

At runtime Widget is an Entity, so it's easier to Bevy to deal with. At project time it's a Scene (or a EntityTree or something like that) so it's easier to user deal with it.

> How should the user define or declare a widget?

Using scenes, be it editing a file, using editor or generating by code.

> How should the user create instances of the widget when building layouts?

Using composable/nested scenes. Adding a button to an existing UI is just a matter to referencing an existing scene prefab.

> How should widgets be composable?

Using scenes (yeah scenes are my silver bullet) as stated above.

> How should widgets be stylable, themable?

There should be a resource used for global theming. Bevy should provide a default theme and users can also add it's own theme and also be able to change it on runtime. Also it should be possible to pass a custom theme for a custom Widget, when instanciating it (to Scenes on project time). When the Widget is created, it should first check if there is a custom widget, if not, it should look for the global theme settings.

> How should the user hook up communication between widgets?

We should have `Input Widgets`, which are using to get data from final-user, like `Buttons`, `InputText`, `DropList` and anything that is used to collect some data from the final-user.

We should have `Display (Output?) Widgets` which doesn't collect input from user, just display some information, like panels, labels, tooltips, status bar, progress bar, graphs and so on.

For `Input Widgets` data, users must be able to collect it using `Query` to a public component.

For both `Input Widgets` and `Display Widgets` users should be able to react events and change Widgets behavior using commands.

> How should widgets send state to the world?

Using components, events and commands. State that should be visible to user, should use public components. State that shouldn't be visible to user, must use some kind of meta components, which store it's internal state and allow for a great flexibility.

> How should widgets request state from the world?

Using systems and querying other widgets public components, events or sending commands.

> Are there other important questions that should be added to this list?

I think each Widget should have it's own Plugin which a set of systems/components, this make it easier to reason about what each widget does. Those plugins should careful expose components, events and commands to interact with other Widgets.

I don't know if I was clear enought, but in short terms we should use Scenes to instanciate, configure and nest/group Widgets together and use Plugins to determine how those Widgets should behave.

**Comment by alice-i-cecile** [comment]
An additional note: theming should be extensible. You should be able to modify new properties using themes, including user-defined ones.

For example, suppose you wanted to add your own glow types, modelled as an enum component. All of the existing tools should support this gracefully.

**Reply by aevyrie** [reply]
Have we considered using `Handle`s to `Asset`s for style? It would be nice to be able to drop in a custom asset handle for styling. This can be done with generic systems. I've been using this technique with mod_picking to allow users to re-use the highlighting logic by adding a `CustomHighlight<T: Asset>` plugin to their app - `StandardMaterial` and `ColorMaterial` versions of this plugin are added by default. We could then use the asset server for hot reloading styles.

Another reason to separate style from widget code is to allow artists to tweak styles without needing help from engineers.

**Comment by Barugon** [comment]
First, I can't even begin to express how happy it makes me that this is finally getting attention. 

> * Immediate mode: The UI is cleared and constructed anew every single frame.
> * Retained mode: The UI is only redrawn when something changes.

Retained mode is obviously going to make layout easier and probably faster but there are some really good aspects to immediate mode, like events that don't require some form of callback mechanism. Perhaps there could be a hybrid approach?

**Reply by Barugon** [reply]
Even egui has state for many of it's widgets but they attempt to hide it from you, which I think is a mistake.

**Comment by cart** [comment]
> Do you agree with the requirements outlined above? 

Yup! I don't see anything particularly controversial in there and I suspect we will find consensus quickly on these high level requirements.

> What should the concrete definition of a Widget be, in terms of entities, components, systems and events?

I think there are two paths here (which are compatible):

1. A hierarchical (parent / child) set of entities and components with a single shared root entity. We should take this path first
2. Logic that generates (1) according to data inputs (generally passed in from a parent) and regenerates them when the inputs change. (React-like). We shouldn't pick this up until we've sorted out (1) and have created a productive workflow there.

> How should the user define or declare a widget?

Open question worth exploring. We can probably use whatever ultimately fills the "Bevy Prefabs / Scenes" niche for this and that should be assumed to be our default approach (at least until we start exploring reactivity, which should happen later).

We can assume that the Prefab / Scene niche will have:
* Scene nesting (in our case, widget nesting)
* Property overloading for nested scenes (in our case, overwriting the default values of components in a widget)
* Asset Handles (in our case, the ability to reference things like font and theme assets)
* A reasonable human readable / editable format 
* A reasonable code-driven definition and instantiation API

> How should the user create instances of the widget when building layouts?

By default, I think we should try to use whatever patterns we use elsewhere in Bevy for prefab / scene spawning (Commands or an interface on top of them)

> How should widgets be composable?

Hierarchical nesting

> How should widgets be stylable, themable?

Open question worth exploring. Setting fields on components which drive specific logic in the widget makes the most sense to me as a starting point. Aka "normal bevy entity configuration".

Notably, this could pair with Reflect based property setting / patching.

Consider the following "real" instance of an Entity hierarchy in a world with Node and Children components:

```
Entity1
    Node
        background: BLUE
        width: 100px
        height: 200px
    Children
        Entity2
            Node
                background: RED
                width: 50%
                height: 40%
```

This could be "themed" by applying a scene-like Reflect based property patch:

```
Entity
    Node
        background: GREEN
        width: 200px
    Children
        Entity
            Node
                height: 30%
```

Scenes will ultimately allow you to omit default fields and just define properties that have changed (important for both compressing scene sizes and seamlessly supporting new fields + picking up new defaults). That "scene field overloading logic" seems very similar to the logic outlined above. We might be able to reuse logic.

> How should the user hook up communication between widgets?

_Very_ open question. Likely some combination of events and direct component value access (via systems). Note that for something "reactive" (again ... we should tackle this later) controlling information flow is critical. As much as possible, communication should only happen in a "top down" fashion by passing data from the parent into the child. ("pure" components in React terms).

But yeah dataflow is _the_ question for UI and not something we can resolve here. People should start thinking about this, developing detailed proposals, and making their case in new Discussions or RFCs. 

> How should widgets send state to the world?

Imo this is the same question as the "widget communication" problem, provided that we agree widgets are just entity / component hierarchies.

> How should widgets request state from the world?

Imo this is the same question as the "widget communication" problem, provided that we agree widgets are just entity / component hierarchies.

> Are there other important questions that should be added to this list?

I think we've already bitten off a lot of scope in this discussion and we shouldn't expand it much further. It looks like we've established relative consensus on the high level stuff. So it makes sense to break out into more scoped conversations for specific topics. Ex: we should have a discussion for event handling, a discussion for widget definition apis, etc. Imo people with strong opinions should start spec-ing out how these should look and behave in practice.

One semi-important question is how we handle "collections / lists". If you want an "inventory widget", how do you keep that in sync with a list of items in the inventory? In some ways, this is just an extension of the dataflow problem, but it is one of the "hard" problems in UI framework design and worth addressing specifically when solving the dataflow problem.

**Reply by CooCooCaCha** [reply]
Regarding point #2 about a react-like system. I've been toying with the idea of a reactive system as a generic logic layer on top of the ECS.

Recently I played around with a 3d rendering engine built in react and it was kinda nice for organizing high-level logic. This could be an alternative to a scripting language as a way to write that kind of logic. The major downside of course is that you introduce a second way to write logic in Bevy that is less performant than plain ECS logic.

Here's an example in the form of a react-like component:
```javascript
function Game() {
  const [isPlaying, setIsPlaying] = useState(false)

  const startGame = () => setIsPlaying(true)

  return (
    <>
      {!isPlaying && <MainMenu onStartGame={startGame} />}
      {isPlaying && (
        <GameSession>
          <Player />
          <Level />
          <Enemies />
        </GameSession>
      )}
    </>
  )
}
```

**Reply by CooCooCaCha** [reply]
Another upside would be that such a system would make the UI more integrated into the rest of the engine. Instead of "this is how you write bevy code and this is how you write UI code."

**Reply by StarArawn** [reply]
@CooCooCaCha This is essentially what Kayak UI does. :)
https://github.com/StarArawn/kayak_ui

One thing to point out though is that you can't just create state out of thin air. In Kayak we create state by spawning an entity that is attached to a widget. To change state you just query the state data as &mut.

Kayak creates reactivity by splitting widgets into two separate systems.
1. Update System - Update system is responsible for handling the reactivity properly. This means diffing data coming into the widget. This data can be coming from "props" IE data passed from a parent widget, or whatever else the user decides(state, context, bevy resource, bevy query).
2. Render System - Responsible for using the incoming data to spawn/update widgets and will only run if the result of the diff there is a change. There is an RSX(like react's jsx) proc macro that handles converting XML like structure into direct calls to bevy. This is optional and you can just spawn widgets using the normal bevy flow. The tree updates/diff are also handled at the end of each render system. The update/diff only apply to a small subset of the tree.

This does have some drawbacks:
1. Widget systems can't run in parallel, they need to follow the order of the tree.
2. Widget diffing needs to happen every frame. This is because really any part of the ECS world can trigger a re-render. I think the cost here is rather small in practice, but it does depend on the size of the UI.

Overall I'm pretty happy with the direction that Kayak is heading and I think it opens up some possible ideas for bevy_ui.
